suite/tests/linux/threadset.dll.c

```diff
@@ -0,0 +1,86 @@
+/* **********************************************************
+ * Copyright (c) 2024 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* This client helpfully instruments every basic block, and flushes
+ * each block with 4% probability! It is great at causing synchalls.
+ */
+
+#include "dr_api.h"
+
+static void
+bb_event(void *p)
+{
+    void *drcontext = dr_get_current_drcontext();
+    static int count = 0;
+
+    /* Avoids executing the hook twice after redirecting execution */
+    if (dr_get_tls_field(drcontext) != NULL) {
+        dr_set_tls_field(drcontext, (void *)0);
+        return;
+    }
+
+    if (++count % 25 == 0) {
+        dr_flush_region(p, 1);
+        /* if we don't sleep, we will interrupt the other thread
+         * too quickly, hitting the (count++ > 3) assert in os.c
```

Comment:
Hmm, is this going to make this test flaky, with some timings on slow automated test suite VM's hitting that assert?  Should we downgrade that assert to a non-fatal ASSERT_CURIOSITY which would still alert us if it happens somewhere we don't expect, but won't fail a test?